src/main/cpp/src/query_operations/broad_combined_gvcf.cc

```diff
@@ -304,28 +306,53 @@ void BroadCombinedGVCFOperator::handle_FORMAT_fields(const Variant& variant)
         ? m_remapped_variant : variant;
     auto valid_field_found = m_field_handlers[variant_type_enum]->collect_and_extend_fields(src_variant, *m_query_config,
         query_field_idx, &ptr, num_elements,
-        m_use_missing_values_not_vector_end && !is_char_type, m_use_missing_values_not_vector_end && is_char_type);
+        m_use_missing_values_not_vector_end && !is_char_type, m_use_missing_values_not_vector_end && is_char_type,
+        known_field_enum == GVCF_GT_IDX);
     if(valid_field_found)
     {
       auto j=0u;
       auto do_insert = true;    //by default, insert into VCF record
       switch(known_field_enum)
       {
         case GVCF_GT_IDX: //GT field is a pita
-          int_vec = const_cast<int*>(reinterpret_cast<const int*>(ptr));
-          //GATK CombineGVCF does not produce GT field by default - option to produce GT
-          if(m_vcf_adapter->produce_GT_field())
           {
-            for(j=0u;j<num_elements;++j)
-              int_vec[j] = is_bcf_valid_value(int_vec[j]) ? bcf_gt_unphased(int_vec[j]) : bcf_gt_missing;
-          }
-          else
-          {
-            //CombineGVCF sets GT field to missing
-            for(j=0u;j<num_elements;++j)
-              int_vec[j] = bcf_gt_missing;
+            int_vec = const_cast<int*>(reinterpret_cast<const int*>(ptr));
+            //GT of type 0/2 is stored as [0,0,2] in GenomicsDB if phased ploidy is used, else [0, 2]
+            //GT of type 0|2 is stored as [0,1,2] in GenomicsDB if phased ploidy is used, else [0, 2]
+            auto GT_length_descriptor = m_query_config->get_length_descriptor_for_query_attribute_idx(m_GT_query_idx);
+            auto num_elements_per_sample = num_elements/variant.get_num_calls();
+            auto store_phase_information_in_GT = GT_length_descriptor == BCF_VL_Phased_Ploidy;
+            auto max_ploidy = KnownFieldInfo::get_ploidy(GT_length_descriptor, num_elements_per_sample);
+            //skip over phasing elements
+            auto step = store_phase_information_in_GT ? 2u : 1u;
+            auto output_idx = 0ull;
+            //GATK CombineGVCF does not produce GT field by default - option to produce GT
+            if(m_vcf_adapter->produce_GT_field())
```

Comment:
this block probably can be simplified and moved into a separate function

Reply:
Any suggestions how to achieve this? I dislike this code snippet as well but couldn't find a better way.
The code is so specific to the output format that it's almost impossible to re-use.

Reply:
I was thinking more on these lines:
line 311: if(!valid_field_found) continue;
remove 313
replace inner loop as
auto x=0u;
auto output_idx=1ull;
for(auto j=0ull;j<num_elements;j+=num_elements_per_sample) {
  v=int_vec[j];
  int_vec[output_idx++] = is_bcf_valid_value(v) ? ((store_phase_information_in_GT) ? bcf_gt_phased(v) : bcf_gt_unphased(v)) : v;
                for(auto k=1u;k<num_elements_per_sample;k+=step) {
                    x=j+k;
                    v=int_vec[x];
                  int_vec[output_idx++] = is_bcf_valid_value(v) ? ((store_phase_information_in_GT && int_vec[x-1])
                      ? bcf_gt_phased(v) : bcf_gt_unphased(v)) : v;
                    }
              }

We can also push the nested for-loop into another function and handle the flag to whether produce_GT_field() or not in there

Reply:
Fixed - if-else statements are in the initialization phase only